From 06774d8899c74929b1559d78fd565f589be51073 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Mon, 26 May 2025 11:24:22 -0400
Subject: [PATCH 867/954] fix(st-ddr): improve error management during PHY init
 on STM32MP1

stm32mp1_ddrphy_idone_wait() returns a negative status if any error
occurs in PGSR during loop.
Parse ZQ0SR0 done/err bit fields after calling stm32mp1_ddrphy_init().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 34384dcbe659e4ff68289615fd476a738c6362db fix(st-ddr): improve error management during PHY init on STM32MP1]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I7536910c616580c488ec5574b3557d648f4fdb71
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/461969
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/ddr/stm32mp1_ddr.c          | 48 ++++++++++++++++++++------
 include/drivers/st/stm32mp1_ddr_regs.h |  5 ++-
 2 files changed, 42 insertions(+), 11 deletions(-)

diff --git a/drivers/st/ddr/stm32mp1_ddr.c b/drivers/st/ddr/stm32mp1_ddr.c
index 51d47507cf..a1fced865d 100644
--- a/drivers/st/ddr/stm32mp1_ddr.c
+++ b/drivers/st/ddr/stm32mp1_ddr.c
@@ -209,7 +209,7 @@ static const struct stm32mp_ddr_reg_info ddr_registers[REG_TYPE_NB] = {
 	},
 };
 
-static void stm32mp1_ddrphy_idone_wait(struct stm32mp_ddrphy *phy)
+static int stm32mp1_ddrphy_idone_wait(struct stm32mp_ddrphy *phy)
 {
 	uint32_t pgsr;
 	int error = 0;
@@ -252,11 +252,15 @@ static void stm32mp1_ddrphy_idone_wait(struct stm32mp_ddrphy *phy)
 	} while (((pgsr & DDRPHYC_PGSR_IDONE) == 0U) && (error == 0));
 	VERBOSE("\n[0x%lx] pgsr = 0x%x\n",
 		(uintptr_t)&phy->pgsr, pgsr);
+
+	return -error;
 }
 
-static void stm32mp1_ddrphy_init(struct stm32mp_ddrphy *phy, uint32_t pir)
+static int stm32mp1_ddrphy_init(struct stm32mp_ddrphy *phy, uint32_t pir)
 {
 	uint32_t pir_init = pir | DDRPHYC_PIR_INIT;
+	uint32_t zq0sr0;
+	int ret = 0;
 
 	mmio_write_32((uintptr_t)&phy->pir, pir_init);
 	VERBOSE("[0x%lx] pir = 0x%x -> 0x%x\n",
@@ -267,7 +271,17 @@ static void stm32mp1_ddrphy_init(struct stm32mp_ddrphy *phy, uint32_t pir)
 	udelay(DDR_DELAY_10US);
 
 	/* Wait DRAM initialization and Gate Training Evaluation complete */
-	stm32mp1_ddrphy_idone_wait(phy);
+	ret = stm32mp1_ddrphy_idone_wait(phy);
+	if (ret != 0) {
+		return ret;
+	}
+
+	zq0sr0 = mmio_read_32((uintptr_t)&phy->zq0sr0);
+	if (((zq0sr0 & DDRPHYC_ZQ0SRN_ZDONE) == 0U) || ((zq0sr0 & DDRPHYC_ZQ0SRN_ZERR) != 0U)) {
+		ret = -EINVAL;
+	}
+
+	return ret;
 }
 
 /* Wait quasi dynamic register update */
@@ -623,7 +637,7 @@ void stm32mp1_refresh_compensation(const struct stm32mp_ddr_config *config,
 	}
 }
 
-static void stm32mp1_self_refresh_zcal(struct stm32mp_ddr_priv *priv, uint32_t zdata)
+static int stm32mp1_self_refresh_zcal(struct stm32mp_ddr_priv *priv, uint32_t zdata)
 {
 	/* sequence for PUBL I/O Data Retention during Power-Down */
 
@@ -652,7 +666,7 @@ static void stm32mp1_self_refresh_zcal(struct stm32mp_ddr_priv *priv, uint32_t z
 	/* 14. Wait for ZQ calibration to finish by polling a 1 status
 	 * on PGSR.IDONE.
 	 */
-	stm32mp1_ddrphy_init(priv->phy, DDRPHYC_PIR_ZCAL);
+	return stm32mp1_ddrphy_init(priv->phy, DDRPHYC_PIR_ZCAL);
 }
 
 void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
@@ -804,7 +818,10 @@ void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
 	 *  4. Monitor PHY init status by polling PUBL register PGSR.IDONE
 	 *     Perform DDR PHY DRAM initialization and Gate Training Evaluation
 	 */
-	stm32mp1_ddrphy_idone_wait(priv->phy);
+	ret = stm32mp1_ddrphy_idone_wait(priv->phy);
+	if (ret != 0) {
+		panic();
+	}
 
 	/*
 	 *  5. Indicate to PUBL that controller performs SDRAM initialization
@@ -827,10 +844,15 @@ void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
 		pir |= DDRPHYC_PIR_ZCALBYP;
 	}
 
-	stm32mp1_ddrphy_init(priv->phy, pir);
+	ret = stm32mp1_ddrphy_init(priv->phy, pir);
 
 	if (config->self_refresh) {
-		stm32mp1_self_refresh_zcal(priv, config->zdata);
+		/* Re-init PHY with ZQ calibration */
+		ret = stm32mp1_self_refresh_zcal(priv, config->zdata);
+	}
+
+	if (ret != 0) {
+		panic();
 	}
 
 	/*
@@ -894,10 +916,16 @@ void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
 		pir |= DDRPHYC_PIR_RVTRN;
 	}
 
-	stm32mp1_ddrphy_init(priv->phy, pir);
+	ret = stm32mp1_ddrphy_init(priv->phy, pir);
+	if (ret != 0) {
+		panic();
+	}
 
 	/* 11. monitor PUB PGSR.IDONE to poll completion of training sequence */
-	stm32mp1_ddrphy_idone_wait(priv->phy);
+	ret = stm32mp1_ddrphy_idone_wait(priv->phy);
+	if (ret != 0) {
+		panic();
+	}
 
 	/* Refresh compensation: forcing refresh command */
 	if (config->self_refresh) {
diff --git a/include/drivers/st/stm32mp1_ddr_regs.h b/include/drivers/st/stm32mp1_ddr_regs.h
index c3ab10c161..9880de215f 100644
--- a/include/drivers/st/stm32mp1_ddr_regs.h
+++ b/include/drivers/st/stm32mp1_ddr_regs.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2017-2025, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: GPL-2.0+ OR BSD-3-Clause
  */
@@ -188,6 +188,9 @@ struct stm32mp_ddrphy {
 #define DDRPHYC_ZQ0CRN_ZDEN			BIT(28)
 #define DDRPHYC_ZQ0CRN_ZQPD			BIT(31)
 
+#define DDRPHYC_ZQ0SRN_ZERR			BIT(30)
+#define DDRPHYC_ZQ0SRN_ZDONE			BIT(31)
+
 #define DDRPHYC_DXNGCR_DXEN			BIT(0)
 
 #define DDRPHYC_DXNDLLCR_DLLSRST		BIT(30)
-- 
2.53.0

